builtin/providers/azurerm/resource_arm_virtual_machine_scale_set.go

```diff
@@ -1037,33 +1087,53 @@ func expandAzureRMVirtualMachineScaleSetsStorageProfileOsDisk(d *schema.Resource
 	osDiskConfig := osDiskConfigs[0].(map[string]interface{})
 	name := osDiskConfig["name"].(string)
 	image := osDiskConfig["image"].(string)
+	vhd_containers := osDiskConfig["vhd_containers"].(*schema.Set).List()
 	caching := osDiskConfig["caching"].(string)
 	osType := osDiskConfig["os_type"].(string)
 	createOption := osDiskConfig["create_option"].(string)
+	managedDiskType := osDiskConfig["managed_disk_type"].(string)
 
 	osDisk := &compute.VirtualMachineScaleSetOSDisk{
-		Name:         &name,
 		Caching:      compute.CachingTypes(caching),
 		OsType:       compute.OperatingSystemTypes(osType),
 		CreateOption: compute.DiskCreateOptionTypes(createOption),
 	}
 
 	if image != "" {
+		osDisk.Name = &name
 		osDisk.Image = &compute.VirtualHardDisk{
 			URI: &image,
 		}
-	} else {
+	}
+
+	if len(vhd_containers) > 0 {
 		var vhdContainers []string
-		containers := osDiskConfig["vhd_containers"].(*schema.Set).List()
-		for _, v := range containers {
+		for _, v := range vhd_containers {
 			str := v.(string)
 			vhdContainers = append(vhdContainers, str)
 		}
+		osDisk.Name = &name
```

Comment:
@tombuildsstuff @sozercan @echuvyrov IMHO, I think it's best in this case to validate that the name is nil/empty in the terraform template (give an error if not), then set it to null when passing to Azure.  Otherwise you'll have problems if Azure later changes and allows this parameter and you start passing in what were previously ignored and likely unintended/invalid values.